### PR TITLE
Feat/misc/notifications/integrate-user-vm : Integrate profile VM to notification interface

### PR DESCRIPTION
**Pull request by charliemangano**
# Integrate profile VM to notification interface

## Description
This PR integrated the user VM into the push notification interface, so that we now can push FCM tokens to our repository. It closes issue #267 and part of #232.

## Changes
This essentially adds a `UserViewModel` parameter to the `PushNotificationsServiceImpl` class and implements a method `uploadToken` to push said token to the profile of the user. To do that, I had to add an attribute `fcmToken`/`fcm_token` to the data classes `User` and `UserDto`, and update the user model to take this new attribute into account when manipulating a profile.

I also updated the authentication screens `SignIn` and `SignUp` to create an FCM token upon login/signing up.

This was all very complicated since the constructors of `PushNotificationsServiceImpl` are restricted by the manifest (see [here](https://github.com/PeriodPals/periodpals/pull/234#discussion_r1855455693) for more explanations) and I was unable to create a user VM factory. This solution is the best I could come up with.

#### Added
None
#### Modified
- `app/src/main/java/com/android/periodpals/services/PushNotificationsServiceImpl.kt`: integrated user VM and implemented method to push FCM token to repository
- `app/src/main/java/com/android/periodpals/model/user/User.kt` and `app/src/main/java/com/android/periodpals/model/user/UserDto.kt`: updated data classes to have FCM token attribute
- `app/src/main/java/com/android/periodpals/MainActivity.kt`: called `PushNotificationsServiceImpl` with proper parameters
- `app/src/main/java/com/android/periodpals/ui/authentication/SignIn.kt` and `app/src/main/java/com/android/periodpals/ui/authentication/SignUp.kt`: updated to create an FCM token upon login/signing up
- `app/src/test/java/com/android/periodpals/services/PushNotificationsServiceImplTest.kt`: added tests for uploading token to repository
- `app/src/test/java/com/android/periodpals/model/user/UserModelSupabaseTest.kt` and `app/src/test/java/com/android/periodpals/model/user/UserViewModelTest.kt`: updated tests to take into account new FCM token attribute in user
#### Removed
None

## Dependencies Added
None

## Testing
Testing this was very complicated as well but in I ended up managing to test `onNewToken` and `createDeviceToken` properly.
I also updated the user model and view model tests to take the new FCM token attribute into account.

### CODE REVIEW

**Review by taghizadlaura — APPROVED** [review]
## Summary
LGTM : 
This PR integrates the `UserViewModel` into the push notification interface, enabling the pushing of FCM tokens to the repository.

## Important (which includes)

### Code Quality
LGTM :
- **Readability**: Clear and well-structured code. Good job maintaining clarity despite the complexity of `PushNotificationsServiceImpl`.
- **Consistency**: Consistent with the project structure.

### Functionality
LGTM :
- **Correctness**: The `uploadToken` method seems to work as intended, pushing FCM tokens to the user profile.

### Performance
LGTM !

### Security 
No security vulnerabilities found.

## Testing
LGTM !
Good job updating the existing tests and adding new ones !

## Documentation
LGTM ! The code changes are well-documented !

## Suggestions
None.

## Steps before PR approved
LGTM !

File: app/src/main/java/com/android/periodpals/model/user/User.kt [hunk 1]
```diff
@@ -8,9 +8,20 @@ package com.android.periodpals.model.user
  * @property description A brief description of the user.
  * @property dob The date of birth of the user.
  */
-data class User(val name: String, val imageUrl: String, val description: String, val dob: String) {
+data class User(
```
Comment (lazarinibruno) [inline app/src/main/java/com/android/periodpals/model/user/User.kt:11]: The documentation for the `fcmToken` is missing.

File: app/src/main/java/com/android/periodpals/model/user/UserDto.kt [hunk 2]
```diff
@@ -15,10 +15,16 @@ data class UserDto(
     val name: String,
     val imageUrl: String,
     val description: String,
-    val dob: String
+    val dob: String,
+    val fcm_token: String? = null,
```
Comment (lazarinibruno) [inline app/src/main/java/com/android/periodpals/model/user/UserDto.kt:19]: - The documentation for the `fcm_token` is missing.
- **Nitpick**: consider renaming `fcm_token` to `fcmToken` to follow Kotlin conventions.

File: app/src/main/java/com/android/periodpals/model/user/UserModelSupabase.kt [hunk 3]
```diff
@@ -18,7 +18,7 @@ class UserRepositorySupabase(private val supabase: SupabaseClient) : UserReposit
 
   override suspend fun loadUserProfile(
       onSuccess: (UserDto) -> Unit,
-      onFailure: (Exception) -> Unit
+      onFailure: (Exception) -> Unit,
```
Comment (lazarinibruno) [inline app/src/main/java/com/android/periodpals/model/user/UserModelSupabase.kt:21]: Good additions having the `onFailure` callback.

File: app/src/main/java/com/android/periodpals/services/PushNotificationsServiceImpl.kt [hunk 4]
```diff
@@ -189,23 +216,26 @@ class PushNotificationsServiceImpl(private val activity: ComponentActivity) :
...
-    firebase.token.addOnCompleteListener { task ->
-      if (task.isSuccessful) {
-        val token = task.result
-        Log.d(TAG, "Token created successfully: ${task.result}")
-        // TODO: send token to server
-      } else {
-        Log.w(TAG, "Failed to get token")
-      }
+  private fun uploadToken(token: String) {
+    if (userViewModel == null) {
```
Comment (lazarinibruno) [inline app/src/main/java/com/android/periodpals/services/PushNotificationsServiceImpl.kt:228]: Good use of control logic to make code more readable.

**Review by lazarinibruno — APPROVED** [review]
# Summary

Many thanks for your hard work! Seems like it was an absolute nightmare implementing this, but you succeeded! It is another solid step towards having working notifications 🔥💪.

## Code quality

As always, code is very well written. It was easy to follow the changes that you implemented.

## Testing

- You updated the existing tests to account for the changes.
  
- The tests that you added to the `PushNotificationServiceImplTest` are thorough, testing different scenarios and what the expected behaviour should be.
  
- All tests pass locally!
  

## Documentation

- You added pertinent comments were needed.
  
- There are some docs missing (see comments), but otherwise good doc.
  

## Steps before merging

Overall very good job. Please do consider my comments. Otherwise, LGTM.

File: app/src/main/java/com/android/periodpals/model/user/User.kt (hunk 1 above)
Comment (charliemangano) [reply app/src/main/java/com/android/periodpals/model/user/User.kt:11]: I'll update, good catch!

File: app/src/main/java/com/android/periodpals/model/user/UserDto.kt (hunk 2 above)
Comment (charliemangano) [reply app/src/main/java/com/android/periodpals/model/user/UserDto.kt:19]: I'll update the docs! 
The name `fcm_token` is to match with the associated column in the Supabase table and the Edge Functions, so I won't touch it

Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/model/user/UserDto.kt:20]: > The name `fcm_token` is to match with the associated column in the Supabase table and the Edge Functions, so I won't touch it

Okay, my bad, I had not understood that.

**Pull request merged**